### PR TITLE
73 label only Carlini

### LINKED ISSUE #91: Handling number of classes
# Issue
## Problem Description
When using ModelHandler(), the init_params is set to null. So if the number of classes is something other than the default value, it causes an error. 
Info regarding init_param  in audit.yaml is either not used or overwritten by the mentioned assignment

### PR DESCRIPTION
**Pull request by fazelehh**
# Description
This PR is ultimately intended to merge with the main but currently, there is another PR ( #89) that should take place first. That is why this is a PR to the result of #89.

Summary of changes

* Renaming Model.py to signal_extractor.py 
* Adding util folder for the signals 
* Adding HSJ distance to the util folder of signals 
* Adding HSJ attack Adding input handler for CIFAR100 
* Fixing loss_traj bug related to iterating a data loader

## Resolved Issues

-  fixes #91 
- 

## Related Pull Requests

- #PR

### CODE REVIEW

File: leakpro/signals/utils/HopSkipJumpDIstance.py [hunk 1]
```diff
@@ -0,0 +1,794 @@
...
+            rv = rv / torch_sqrt(torch_sum(rv ** 2, dim=axis, keepdim=True))
+            rv = rv.to(self.device)
+            perturbed = sample + delta[i] * rv
+            perturbed = clamp(perturbed, self.clip_min, self.clip_max)
+            rv = (perturbed - sample) / delta[i]
+
+            # query the model.
+            sample_index = np.full(num_evals, active_indices[i])
+            decisions, _ = self.decision_function(perturbed, sample_index)
+            fval = 2.0 * decisions.view((decisions.shape[0],) +
```
Comment (johanos1) [inline leakpro/signals/utils/HopSkipJumpDIstance.py:693]: make comment: "+1 if decisions = True, else -1

File: leakpro/signals/utils/HopSkipJumpDIstance.py [hunk 2]
```diff
@@ -0,0 +1,794 @@
...
+
+        """
+        batch_epsilon = dist / np.sqrt(current_iteration)
+        active_indices = np.array(active_indices)
+
+        success = np.zeros(len(samples), dtype=bool)
+        batch_active_indices = np.arange(len(samples))
+        num_evals = 0
+
+        while not success.all() and num_evals <= 10:
```
Comment (johanos1) [inline leakpro/signals/utils/HopSkipJumpDIstance.py:588]: num_evals is hard coded.

Comment (johanos1) [reply leakpro/signals/utils/HopSkipJumpDIstance.py:588]: I think it makes sense to have a num_evals here. It restricts the number of queries we can make to the API. This is a quantity I think the user should be able to set!

File: leakpro/signals/utils/HopSkipJumpDIstance.py [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+            if passed_index:
+                # Find the positions in batch_active_indices corresponding to the values in passed_index
+                positions_to_delete = [i for i, val in enumerate(batch_active_indices) if val in passed_index]
+
+                batch_active_indices = np.delete(batch_active_indices, positions_to_delete)
+            num_evals += 1
+            if self.verbose:
+                message = f"Stepsize search: {num_evals} evaluations, and failed data {len(batch_active_indices)} in {b_i}."
+                self.logger.info(message)
+        return batch_epsilon
```
Comment (johanos1) [inline leakpro/signals/utils/HopSkipJumpDIstance.py:613]: the return does not entail what samples were successful. Isnt this information that is useful to us? Should we really attack a point we didnt succeed with?

File: leakpro/signals/utils/HopSkipJumpDIstance.py [hunk 4]
```diff
@@ -0,0 +1,794 @@
...
+            # search step size.
+            if self.stepsize_search == "geometric_progression":
+                # find step size.
+                epsilon = self.geometric_progression_for_stepsize(
+                    perturbed, update, dist, current_iter, active_indices, b_i
+                )
+
+                # Update the sample.
+                perturbed = clamp(
+                    perturbed + epsilon.view(len(epsilon),1,1,1) * update, self.clip_min, self.clip_max
```
Comment (johanos1) [inline leakpro/signals/utils/HopSkipJumpDIstance.py:298]: Again, not convinced with clamp.

File: leakpro/attacks/mia_attacks/HSJ.py [hunk 5]
```diff
@@ -124,52 +132,66 @@ def run_attack(self:Self) -> CombinedMetricResult:
...
+                                                self.batch_size,
+                                                self.verbose,
+                                                self.clip_min,
+                                                self.clip_max,
                                                 )
-        sanity_check = True
-        if sanity_check:
-            self.sanity_check(perturbation_distances_in, perturbation_distances_out)
+
         np.save("perturbation_in.npy", perturbation_distances_in)
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/HSJ.py:188]: should be stored to attack-specific folder in leakpro output

File: leakpro/attacks/mia_attacks/HSJ.py [hunk 6]
```diff
@@ -124,52 +132,66 @@ def run_attack(self:Self) -> CombinedMetricResult:
...
         np.save("perturbation_out.npy", perturbation_distances_out)
 
 
+        if self.reproducing_paper_results:
+            self.sanity_check(perturbation_distances_in, perturbation_distances_out)
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/HSJ.py:193]: what is this for?

File: leakpro/attacks/mia_attacks/HSJ.py [hunk 7]
```diff
@@ -274,78 +296,218 @@ def sanity_check(self: Self, shadow_distances_in: np.ndarray, shadow_distances_o
...
+        attack_in_indices = np.random.choice(target_train_indices, aux_data_size, replace=False)
+        attack_out_indices = np.random.choice(target_test_indices, aux_data_size, replace=False)
+
+        attack_in_dataset = self.population.subset(attack_in_indices)
+        attack_out_dataset = self.population.subset(attack_out_indices)
+
+        attack_in_dataloader = DataLoader(attack_in_dataset, batch_size=self.batch_size, shuffle=True)
+        attack_out_dataloader = DataLoader(attack_out_dataset, batch_size=self.batch_size, shuffle=True)
+
+        _, in_data_distances = self.signal( self.target_model,
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/HSJ.py:312]: Here, the target is used as input on the actual target data. Why did we use shadow models in the attack?

File: leakpro/attacks/mia_attacks/HSJ.py [hunk 8]
```diff
@@ -274,78 +296,218 @@ def sanity_check(self: Self, shadow_distances_in: np.ndarray, shadow_distances_o
...
+                                            self.gamma,
+                                            self.constraint,
+                                            self.batch_size,
+                                            self.verbose,
+                                            self.clip_min,
+                                            self.clip_max,
                                             )
 
+        # Find the optimal threshold
+        self.best_threshold_precision(self.find_threshold1(in_data_distances, out_data_distances),
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/HSJ.py:348]: what are all these?

File: leakpro/attacks/mia_attacks/HSJ.py [hunk 9]
```diff
@@ -124,52 +132,66 @@ def run_attack(self:Self) -> CombinedMetricResult:
...
-        if sanity_check:
-            self.sanity_check(perturbation_distances_in, perturbation_distances_out)
+
         np.save("perturbation_in.npy", perturbation_distances_in)
         np.save("perturbation_out.npy", perturbation_distances_out)
 
 
+        if self.reproducing_paper_results:
+            self.sanity_check(perturbation_distances_in, perturbation_distances_out)
+
```
Comment (johanos1) [inline leakpro/attacks/mia_attacks/HSJ.py:194]: Why is the normalizations on line 208 needed? They dont do anything?

Comment (johanos1) [reply leakpro/attacks/mia_attacks/HSJ.py:194]: I think the thresholds are off. You create the thresholds based on the non-transformed signals but then you use the transformed signals for thresholding. Why do we need to even transform the signals if it is done this way?

**Review by johanos1 — CHANGES_REQUESTED** [review]

**Review by johanos1 — APPROVED** [review]
